### PR TITLE
Support buffering for streams on both sides.

### LINKED ISSUE #28: Session Establishment Race
> From the client's perspective, a WebTransport session is established when the client receives a 200 response. From the server's perspective, a session is established once it sends a 200 response. Both endpoints MUST NOT open any streams or send any datagrams on a given session before that session is established.

After the client initiates the WebTransport session by sending a CONNECT request, the server sends a 200 response and is free to create streams. This means that a server-initiated stream may arrive prior to the CONNECT response. Is the client able to infer that the session was successfully established, or does the client need to buffer this stream until it receives a proper response?

### PR DESCRIPTION
**Pull request by vasilvv**
Fixes #28, #29.

### CODE REVIEW

**Comment by vasilvv** [comment]
cc @afrind @kixelated

**Comment by kixelated** [comment]
Looks great!

File: draft-ietf-webtrans-http3.md [hunk 1]
```diff
@@ -323,6 +322,25 @@ However, when proxying the datagrams, the hop-by-hop MTUs can vary.
 TODO: Describe how the path MTU can be computed, specifically propagation across
 HTTP proxies.
 
+## Buffering Incoming Streams and Datagrams
+
+In WebTransport over HTTP/3, the client MAY send its SETTINGS frame, a
+WebTransport CONNECT request, a WebTransport data stream and a WebTransport
+datagram all within a single flight.  As those can arrive out of order, a
```
Comment (afrind) [inline draft-ietf-webtrans-http3.md:329]: This makes it sounds like the limit is 1 data stream and 1 datagram.  I think you want to say the client can send any number of any type.

Comment (vasilvv) [reply draft-ietf-webtrans-http3.md]: Good point. Fixed.

File: draft-ietf-webtrans-http3.md [hunk 2]
```diff
@@ -323,6 +322,25 @@ However, when proxying the datagrams, the hop-by-hop MTUs can vary.
 TODO: Describe how the path MTU can be computed, specifically propagation across
 HTTP proxies.
 
+## Buffering Incoming Streams and Datagrams
+
+In WebTransport over HTTP/3, the client MAY send its SETTINGS frame, as well as
+multiple WebTransport CONNECT requests, WebTransport data streams and
+WebTransport datagrams, all within a single flight.  As those can arrive out of
+order, a WebTransport server could be put into a situation where it has a
```
[the remark this answers is not in the document]
Comment (vasilvv) [reply draft-ietf-webtrans-http3.md]: Fixed.

**Review by afrind — APPROVED** [review]

File: draft-ietf-webtrans-http3.md [hunk 3]
```diff
@@ -323,6 +322,25 @@ However, when proxying the datagrams, the hop-by-hop MTUs can vary.
 TODO: Describe how the path MTU can be computed, specifically propagation across
 HTTP proxies.
 
+## Buffering Incoming Streams and Datagrams
+
+In WebTransport over HTTP/3, the client MAY send its SETTINGS frame, as well as
+multiple WebTransport CONNECT requests, WebTransport data streams and
+WebTransport datagrams, all within a single flight.  As those can arrive out of
+order, a WebTransport server could be put into a situation where it receives a
+stream or a datagram without a corrseponding session.  Similarly, a client may
```
Comment (DavidSchinazi) [inline draft-ietf-webtrans-http3.md:331]:
```suggestion
stream or a datagram without a corresponding session.  Similarly, a client may
```

**Review by DavidSchinazi — APPROVED** [review]

**Pull request merged**